File: app/src/main/java/com/mkulesh/onpc/iscp/messages/GoogleCastAnalyticsMsg.java
/*
 * Enhanced Controller for Onkyo and Pioneer
 * Copyright (C) 2018-2021 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

package com.mkulesh.onpc.iscp.messages;

import com.mkulesh.onpc.R;
import com.mkulesh.onpc.iscp.EISCPMessage;
import com.mkulesh.onpc.iscp.ISCPMessage;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;

/*
 * Google Cast Share Usage Data Command
 */
public class GoogleCastAnalyticsMsg extends ISCPMessage
{
    public final static String CODE = "NGU";

    public enum Status implements StringParameterIf
    {
        NONE("N/A", R.string.device_two_way_switch_none),
        OFF("00", R.string.device_two_way_switch_off),
        ON("01", R.string.device_two_way_switch_on);

        final String code;

        @StringRes
        final int descriptionId;

        Status(final String code, @StringRes final int descriptionId)
        {
            this.code = code;
            this.descriptionId = descriptionId;
        }

        public String getCode()
        {
            return code;
        }

        @StringRes
        public int getDescriptionId()
        {
            return descriptionId;
        }
    }

    private final Status status;

    GoogleCastAnalyticsMsg(EISCPMessage raw) throws Exception
    {
        super(raw);
        status = (Status) searchParameter(data, Status.values(), Status.NONE);
    }

    public GoogleCastAnalyticsMsg(Status level)
    {
        super(0, null);
        this.status = level;
    }

    public Status getStatus()
    {
        return status;
    }

    @NonNull
    @Override
    public String toString()
    {
        return CODE + "[" + status.toString() + "]";
    }

    @Override
    public EISCPMessage getCmdMsg()
    {
        return new EISCPMessage(CODE, status.getCode());
    }

    @Override
    public boolean hasImpactOnMediaList()
    {
        return false;
    }

    public static Status toggle(Status s)
    {
        return (s == Status.OFF) ? Status.ON : Status.OFF;
    }
}

File: lib/iscp/messages/GoogleCastAnalyticsMsg.dart
/*
 * Enhanced Music Controller
 * Copyright (C) 2019-2023 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */
// @dart=2.9
import "../../constants/Strings.dart";
import "../EISCPMessage.dart";
import "EnumParameterMsg.dart";

enum GoogleCastAnalytics
{
    NONE,
    OFF,
    ON
}

/*
 * Google Cast Share Usage Data Command
 */
class GoogleCastAnalyticsMsg extends EnumParameterMsg<GoogleCastAnalytics>
{
    static const String CODE = "NGU";

    static const ExtEnum<GoogleCastAnalytics> ValueEnum = ExtEnum<GoogleCastAnalytics>([
        EnumItem.code(GoogleCastAnalytics.NONE, "N/A",
            descrList: Strings.l_device_two_way_switch_none, defValue: true),
        EnumItem.code(GoogleCastAnalytics.OFF, "00",
            descrList: Strings.l_device_two_way_switch_off),
        EnumItem.code(GoogleCastAnalytics.ON, "01",
            descrList: Strings.l_device_two_way_switch_on)
    ]);

    GoogleCastAnalyticsMsg(EISCPMessage raw) : super(CODE, raw, ValueEnum);

    GoogleCastAnalyticsMsg.output(GoogleCastAnalytics v) : super.output(CODE, v, ValueEnum);

    @override
    bool hasImpactOnMediaList()
    {
        return false;
    }

    static GoogleCastAnalytics toggle(GoogleCastAnalytics s)
    {
        return (s == GoogleCastAnalytics.OFF) ? GoogleCastAnalytics.ON : GoogleCastAnalytics.OFF;
    }
}
